src/Pointer.huff

```diff
@@ -42,8 +45,15 @@
 ///      allocated to host. Everything before will be owned by the VM, and
 ///      innaccessible by the host.
 #define macro FIX_MEMOFFSET() = takes (1) returns (1) {
-    // Takes                // [mem_offset]
-    [HOST_MEMORY_START] add // [mem_offset + HOST_MEMORY_START]
+    // Takes                        // [mem_offset]
+    dup1                            // [mem_offset, mem_offset]
+    [MAX_MEMORY_OFFSET]             // [MAX_MEMORY_OFFSET, mem_offset, mem_offset]
+    gt                              // [MAX_MEMORY_OFFSET > mem_offset, mem_offset]
+    fix_memoffset jumpi             // [mem_offset]
+    0x00 dup1 revert
+
+    fix_memoffset:
+            [HOST_MEMORY_START] add // [HOST_MEMORY_START + mem_offset]
```

Comment:
this is way more gas efficient AND it will have the same behaviour as the EVM (i.e. `out of gas` when trying to access high memory offsets ... not `revert` like here):

```
    // Takes                // [mem_offset]
    // we'll mask upper bytes to ensure that the addition will not overflow.
    // this will in practice corrupt all the memory accesses to very high addresses
   //  but those cannot be legitimate: they will end-up in out-of-gas anyway.
    0xffffffffffffffffffffffffffffffffffffffff and
    [HOST_MEMORY_START] add // [mem_offset + HOST_MEMORY_START]
```

Reply:
It should be better now